### PR TITLE
Sidecar refactor, reject assertions, flush validators

### PR DESCRIPTION
**Pull request by macwis**
# Goals of PR

Core changes:

- Light refactor (Task-128):
  - App initialization swapped to dynamic code generation with Wire 
  - Initial set of basic providers for Wire
  - Config patching for extracting SystemConfig object from the CLI library context
  - Go linter fixes
  - Forge fmt check fixes
- Reject assertions (Task-157)
- Flush validators (Task-29)

Notes:

For the future: configuration could use some flattening Config with SystemConfig, and others. In concequence the logger level will be unified.

### CODE REVIEW

File: services/sidecar/rollup/rpc/bridge/bridge_client.go [hunk 1]
```diff
@@ -62,6 +63,14 @@ func (c *BridgeClient) GetRequiredStakeAmount(ctx context.Context) (*big.Int, er
 	return c.IRollup.CurrentRequiredStake(&bind.CallOpts{Pending: false, Context: ctx})
 }
 
+func (c *BridgeClient) RequireFirstUnresolvedAssertionIsRejectable(ctx context.Context, address common.Address) error {
+	return c.IRollup.RequireFirstUnresolvedAssertionIsRejectable(&bind.CallOpts{Pending: false, Context: ctx}, address)
+}
+
+func (c *BridgeClient) RejectFirstUnresolvedAssertion(ctx context.Context, address common.Address) (*types.Transaction, error) {
+	return c.IRollup.RejectFirstUnresolvedAssertion(&bind.TransactOpts{Context: ctx}, address)
+}
```
Comment (ujvl) [inline services/sidecar/rollup/rpc/bridge/bridge_client.go:72]: use `TxManager` instead for tx submission (this struct should only be used for `view` functions) to handle reorgs/resubmission etc

File: services/sidecar/rollup/services/validator/validator.go [hunk 2]
```diff
@@ -97,6 +98,10 @@ func (v *Validator) step(ctx context.Context) error {
 	}
 	// TODO: validate assertions locally.
 	// Resolve the first unresolved assertion.
+	// FIXME: THIS IS NOT NECESSARILY THE ASSERTION THAT IS CREATED ABOVE,
+	//        probably we should put that as a separate running services (loops), async
+	// TODO: maybe there is more then one to resolve, so we should maybe try to resolve all available
+	// 		 (take a look at the rollup contract).
```
Comment (ujvl) [inline services/sidecar/rollup/services/validator/validator.go:104]: yes, we can separate this functionality out

File: services/sidecar/rollup/services/validator/validator.go [hunk 3]
```diff
@@ -146,10 +172,105 @@ func (v *Validator) resolveFirstUnresolvedAssertion(ctx context.Context) error {
...
+		if err != nil {
+			// It should be rejectable, but it failed to reject, so it must be a bigger problem
+			// TODO: confirm solution
+			return err
+		}
 	}
+
+	// At this point we know it's confirmable
+	// But first we need to validate
+	// TODO: this is not implemented properly, fill-in with a stub/mock
```
Comment (ujvl) [inline services/sidecar/rollup/services/validator/validator.go:199]: not necessary (they're orthogonal in some sense)

**Review by siosw — APPROVED** [review]

File: services/sidecar/.gitignore [hunk 4]
```diff
@@ -1,3 +1,4 @@
 build
 bindings/I*
 bindings/L1Oracle.go
+*/**/forge-std
```
Comment (LEAFERx) [inline services/sidecar/.gitignore:4]: why this is added?

File: services/sidecar/rollup/rpc/bridge/bridge_client.go [hunk 5]
```diff
@@ -50,3 +51,49 @@ func (c *BridgeClient) GetLastConfirmedAssertionID(ctx context.Context) (*big.In
...
+
+func (c *BridgeClient) RequireFirstUnresolvedAssertionIsRejectable(ctx context.Context, address common.Address) error {
+	return c.IRollup.RequireFirstUnresolvedAssertionIsRejectable(&bind.CallOpts{Pending: false, Context: ctx}, address)
+}
+
+func (c *BridgeClient) IsStakedOnAssertion(ctx context.Context, assertionID *big.Int, address common.Address) (bool, error) {
+	return c.IRollup.IsStakedOnAssertion(&bind.CallOpts{Pending: false, Context: ctx}, assertionID, address)
+}
+
+// Returns the last assertion ID that was validated *by us*.
```
Comment (LEAFERx) [inline services/sidecar/rollup/rpc/bridge/bridge_client.go:63]: any reason to keep these commented out functions?

File: services/sidecar/rollup/rpc/eth/utils.go [hunk 6]
```diff
@@ -37,11 +38,16 @@ func SubscribeNewHeadByPolling(
...
 		}
-		poll()
+		if err := poll(); err != nil {
+			return err
+		}
 		for {
 			select {
 			case <-ticker.C:
-				poll()
+				err := poll()
```
Comment (LEAFERx) [inline services/sidecar/rollup/rpc/eth/utils.go:47]: nit: combine following 2 lines?

File: services/sidecar/rollup/services/disseminator/batch_disseminator.go [hunk 7]
```diff
@@ -80,7 +82,10 @@ func (d *BatchDisseminator) step(ctx context.Context) error {
 	if err := d.appendToBuilder(ctx); err != nil {
 		if errors.As(err, &L2ReorgDetectedError{}) {
 			log.Error("Reorg detected, reverting to safe state.", "error", err)
-			d.rollback()
+			err := d.rollback()
```
Comment (LEAFERx) [inline services/sidecar/rollup/services/disseminator/batch_disseminator.go:85]: nit: combine following 2 lines?

File: services/sidecar/rollup/services/resolver/resolver.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package resolver
```
Comment (LEAFERx) [inline services/sidecar/rollup/services/resolver/resolver.go:1]: is this file unfinished or just a template?

File: services/sidecar/.gitignore (hunk 4 above)
Comment (macwis) [reply services/sidecar/.gitignore:4]: it is trying to include this in the repo, and probably it should not be there, right?

File: services/sidecar/rollup/rpc/bridge/bridge_client.go (hunk 5 above)
Comment (macwis) [reply services/sidecar/rollup/rpc/bridge/bridge_client.go:63]: they were there before, I just moved the code. we can clean that up @ujvl I think you're listed on git blame, is it fine to remove?

File: services/sidecar/rollup/services/resolver/resolver.go (hunk 8 above)
Comment (macwis) [reply services/sidecar/rollup/services/resolver/resolver.go:1]: yes, so I started drafting this, but it shouldn't end up in the commit, must have accidentally add when fixing comments to the PR. will remove. thanks for spotting!

File: services/sidecar/rollup/rpc/bridge/bridge_client.go (hunk 5 above)
Comment (ujvl) [reply services/sidecar/rollup/rpc/bridge/bridge_client.go:63]: yes remove

Comment (macwis) [reply services/sidecar/rollup/rpc/bridge/bridge_client.go]: done

**Review by ujvl — APPROVED** [review]

**Review by ujvl — APPROVED** [review]

**Pull request merged**